internal/monitor/otelexporters.go

```diff
@@ -0,0 +1,119 @@
+// limitations under the License.
+
+package monitor
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/googlecloudplatform/gcsfuse/v2/cfg"
+	"github.com/googlecloudplatform/gcsfuse/v2/common"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/logger"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"go.opentelemetry.io/contrib/detectors/gcp"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/prometheus"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
+)
+
+// SetupOTelMetricExporters sets up the metrics exporters
+func SetupOTelMetricExporters(ctx context.Context, c *cfg.Config) (shutdownFn common.ShutdownFn) {
+	options := make([]metric.Option, 0)
+	opts, shutdownFn := setupPrometheus(c.Metrics.PrometheusPort)
+	options = append(options, opts...)
+	res, err := getResource(ctx, c)
+	if err != nil {
+		logger.Errorf("Error while fetching resource: %v", err)
+	} else {
+		options = append(options, metric.WithResource(res))
+	}
+	meterProvider := metric.NewMeterProvider(options...)
+	otel.SetMeterProvider(meterProvider)
+	return common.JoinShutdownFunc(meterProvider.Shutdown, shutdownFn)
+}
+
+func setupPrometheus(port int64) ([]metric.Option, common.ShutdownFn) {
+	if port <= 0 {
+		return nil, nil
+	}
+	exporter, err := prometheus.New(prometheus.WithoutUnits(), prometheus.WithoutCounterSuffixes(), prometheus.WithoutScopeInfo(), prometheus.WithoutTargetInfo())
+	if err != nil {
+		logger.Errorf("Error while creating prometheus exporter:%v", err)
+		return nil, nil
+	}
+	shutdownCh := make(chan context.Context)
+	done := make(chan interface{})
+	go serveMetrics(port, shutdownCh, done)
+	return []metric.Option{metric.WithReader(exporter)}, func(ctx context.Context) error {
+		shutdownCh <- ctx
+		close(shutdownCh)
+		<-done
+		close(done)
+		return nil
+	}
+}
+
+func serveMetrics(port int64, shutdownCh <-chan context.Context, done chan<- interface{}) {
+	logger.Infof("Serving metrics at localhost:%d/metrics", port)
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.Handler())
+	prometheusServer := &http.Server{
+		Addr:           fmt.Sprintf(":%d", port),
+		Handler:        mux,
+		ReadTimeout:    10 * time.Second,
+		WriteTimeout:   10 * time.Second,
+		MaxHeaderBytes: 1 << 20,
+	}
+	go func() {
+		if err := prometheusServer.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			logger.Errorf("Failed to start Prometheus server: %v", err)
+		}
+	}()
+	go func() {
+		ctx := <-shutdownCh
+		defer func() { done <- true }()
+		logger.Info("Shutting down Prometheus exporter.")
+		if err := prometheusServer.Shutdown(ctx); err != nil {
+			logger.Errorf("Error while shutting down Prometheus exporter:%v", err)
+			return
+		}
+		logger.Info("Prometheus exporter shutdown")
+	}()
+	logger.Info("Prometheus collector exporter started")
+}
+
+func getAppName(c *cfg.Config) string {
```

Comment:
Is there a chance that c.AppName is not populated?
This fallback is better to be there in config.

Reply:
Makes sense - pinned it to gcsfuse.